bpo-27584: New addition of vSockets to the python socket module

caavery opened the pull request:
Support for AF_VSOCK on Linux only

<!-- issue-number: bpo-27584 -->
https://bugs.python.org/issue27584
<!-- /issue-number -->

caavery committed:
bpo-27584: New addition of vSockets to the python socket module

Support for AF_VSOCK on Linux only

caavery commented:
The patch is also in Issue number 27584 along with a README file that explains how to setup the environment for tests.

kushaldas commented on Modules/socketmodule.c:
```diff
@@ -1227,6 +1227,14 @@ makesockaddr(SOCKET_T sockfd, struct sockaddr *addr, size_t addrlen, int proto)
+       case AF_VSOCK:
+       {
+           struct sockaddr_vm *a = (struct sockaddr_vm *) addr;
+           return Py_BuildValue("II", a->svm_cid, a->svm_port);
```
This is the line failing in the Travis. https://travis-ci.org/python/cpython/jobs/248412030#L2063

caavery replied:
You don't have the proper headers. 
checking for linux/vm_sockets.h... no
 
So it looks like my checking for HAVE_LINUX_VM_SOCKETS_H is faulty.

kushaldas replied:
Okay, so please update that, and push in your branch, the bots will pick up the changes and will test.

caavery replied:
I've restored the original aclocal.m4 and fixed the issues you pointed out. 

Thank you for your review,

Cathy

kushaldas commented on aclocal.m4:
```diff
@@ -1,6 +1,6 @@
-# generated automatically by aclocal 1.15 -*- Autoconf -*-
+# generated automatically by aclocal 1.13.4 -*- Autoconf -*-
```
Why are we using an old version here?

caavery replied:
I must have an older version installed. automake-1.13.4-3.el7.noarch I can upgrade that.

So do I make the corrections and ask for another pull?

kushaldas replied:
Just update this PR.

caavery replied:
I rebased my code to remove my patch. Is rebasing and pushing an acceptable way to make changes? Or do I need to add a new commit each time?

Thanks

caavery replied:
Sorry to remove the aclocal.m4 from my patch.

berkerpeksag commented on Lib/test/test_socket.py:
```diff
@@ -75,12 +94,20 @@ def _have_socket_alg():
 
+def _have_socket_vsock():
+    """Check whether AF_VSOCK sockets are supported on this host."""
+    if (getCid() == None):
```
I'd use a shorter version:

```py
return getCid() is not None
```

caavery replied:
is get_cid acceptable?

caavery replied:
Sorry wrong comment.

caavery replied:
I don't understand this syntax. It looks like return the cid if it is not None or do nothing and continue. getCid will return an integer or None. I want to translate that into return True or False in  _have_socket_vsock().

musically-ut replied:
@caavery read @berkerpeksag's suggestion as a contraction of the following:

```python
    foo = getCid is not None
    return foo
```
`foo` will contain a boolean value which will be returned.

caavery replied:
Thanks!

berkerpeksag replied:
Yeah, that's what I was trying to say. Sorry for being unclear!

And +1 for renaming it to `get_cid`, thanks!

berkerpeksag commented on Lib/test/test_socket.py:
```diff
@@ -191,7 +218,6 @@ def setUp(self):
         except OSError:
             self.skipTest('unable to bind RDS socket')
 
-
```
Please don't make unrelated cosmetic changes.

berkerpeksag commented on Lib/test/test_socket.py:
```diff
@@ -5627,6 +5734,10 @@ def test_main():
     tests.append(LinuxKernelCryptoAPI)
     tests.extend([
+        BasicVSOCKTest,
+        ThreadedVSOCKSocketStreamTest
```
Style nit: Please add trailing comma.

berkerpeksag commented on Lib/test/test_socket.py:
```diff
@@ -44,6 +46,23 @@
 
+def getCid():
+    import struct
+    import fcntl
```
You may want to check if fcntl is available on the platform.

berkerpeksag commented on Doc/library/socket.rst:
```diff
@@ -149,6 +149,14 @@ created.  Socket addresses are represented as follows:
   .. versionadded:: 3.6
 
+- :const:`AF_VSOCK` allows communication between virtual machines and
+   their hosts. The sockets are represented as a (CID, port) tuple
```
Indentation is off here:

```rst
- :const:`AF_VSOCK` allows [...]
  their hosts. [...]

  Availability: Linux >= 4.8. QEMU >= 2.8.
```

berkerpeksag replied:
Style nit: (CID, port) -> \`\`(CID, port)\`\`

berkerpeksag commented on Doc/library/socket.rst:
```diff
@@ -149,6 +149,14 @@ created.  Socket addresses are represented as follows:
+   their hosts. The sockets are represented as a (CID, port) tuple
+   where the context ID or CID and port are integers.
+
+  Availability: Linux >= 4.8. QEMU >= 2.8.
```
Do we need to mention QEMU here? I'd say just document kernel version.

caavery replied:
Well I thought it would be better informationally to have it in but if it is not acceptable I can remove it.

berkerpeksag commented on Lib/test/test_socket.py:
```diff
@@ -44,6 +46,23 @@
 except ImportError:
     _socket = None
 
+def getCid():
```
Style nit: Can we use an under_score name here?

caavery replied:
No problem.

caavery replied:
is get_cid acceptable?

berkerpeksag commented on Lib/test/test_socket.py:
```diff
@@ -375,6 +401,40 @@ def clientTearDown(self):
+          "This test can only be run on a virtual guest.")
+class ThreadedVSOCKSocketStreamTest(unittest.TestCase, ThreadableTest):
+
+    def __init__(self, methodName = 'runTest'):
```
Style nit: No need to add spaces around `=`.

```py
methodName= 'runTest'
```

Same for line 411:

```py
methodName=methodName
```

berkerpeksag reviewed (commented):
Thank you for the PR. Could you please add your name to `Misc/ACKS`?

caavery committed:
bpo-27584: Fixes for V2

Fixed syntax and naming problems.
Fixed #ifdef AF_VSOCK checking
Restored original aclocal.m4

caavery committed:
bpo-27584: Fixes for V3

Added checking for fcntl and thread modules.

caavery committed:
bpo-27584: Fixes for V4

Fixed white space error

berkerpeksag commented:
Looks like your branch contains additional commits and unfortunately it makes code review harder. Could you rebase your branch against current master? You might find https://devguide.python.org/gitbootcamp/#syncing-with-upstream helpful.

caavery commented:
@berkerpeksag 

So you want me to sync my github branch bpo-27584 with the upstream cpython branch? Is that correct? I'm just a little unclear as to what you are asking.

Cathy

caavery commented:
@berkerpeksag 

Or do you want me to rebase both my master ( fork of upstream ) and my bpo-27584 branch?

Thanks again

berkerpeksag commented:
> So you want me to sync my github branch bpo-27584 with the upstream cpython branch? Is that correct? 

Correct. Your branch bpo-27584 now contains 34 commits and 30 of them are unrelated to the vSockets feature as you can see at https://github.com/python/cpython/pull/2489/files (the reason is probably [this merge commit](https://github.com/python/cpython/pull/2489/commits/4bc606af2514ab87ebba909f70994798243ef40a))

caavery commented:
@berkerpeksag 

Sorry I seem to have an issue. 
git fetch upstream
git rebase upstream/master

Python Dev~/caavery.github/cpython# git push origin bpo-27584
To git@github.com:caavery/cpython.git
 ! [rejected]        bpo-27584 -> bpo-27584 (non-fast-forward)
error: failed to push some refs to 'git@github.com:caavery/cpython.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Merge the remote changes (e.g. 'git pull')
hint: before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

caavery commented:
@berkerpeksag

Sorry this is my first time with github and working with remote branches. I just forced the push into my branch bpo-27584. My branch says that 'This branch is 4 commits ahead of python:master. ' which is what I think you wanted.

berkerpeksag commented:
> My branch says that 'This branch is 4 commits ahead of python:master. ' which is what I think you wanted.

Looks great, thanks! I will try to take another look at the patch while we wait for Kushal's review.

caavery commented:
@kushaldas 

Is there anything else I should be doing? The patches have passed all tests so far and I have addressed all the issues raised.

Thanks,

Cathy

ilevkivskyi commented on Doc/library/socket.rst:
```diff
@@ -149,6 +149,14 @@ created.  Socket addresses are represented as follows:
   .. versionadded:: 3.6
 
+- :const:`AF_VSOCK` allows communication between virtual machines and
+  their hosts. The sockets are represented as a ``(CID,port)`` tuple
```
Shouldn't there be space after coma? ``(CID, port)``.

caavery replied:
Yes that got accidentally deleted.

caavery committed:
bpo-27584: Fixes for V5

Added back comma in (CID, port).

rmsr commented on Doc/library/socket.rst:
```diff
@@ -149,6 +149,14 @@ created.  Socket addresses are represented as follows:
+  their hosts. The sockets are represented as a ``(CID, port)`` tuple
+  where the context ID or CID and port are integers.
+
+  Availability: Linux >= 4.8. QEMU >= 2.8.
```
Given that vsockets originated with VMWare, add a note that they are available in that environment also?

caavery replied:
This is a bit more complicated. The availability in a particular linux kernel depends on the implementation. 
virtio-vsock is Linux >= 4.8. QEMU >= 2.8.
vmci_transport is Linux >= 3.9
Should I just change it to when it first appeared which is v3.9?

rmsr replied:
I meant availability when running under the VMWare hypervisor. As for the kernel, to keep it simple, whichever version supplies full functionality with both qemu and vmware.

tiran commented on Lib/test/test_socket.py:
```diff
@@ -44,6 +46,22 @@
+def get_cid():
+    if fcntl is None:
+        return None
+    if not os.path.exists("/dev/vsock"):
```
Please rewrite this block with proper context manager. You can skip the os.path.exists check, too.

```
try:
    with open(...) as f:
        r = fcntl.ioctl(f, socket.IOCTL_VM_SOCKETS_GET_LOCAL_CID, "    ")
except OSError:
    return None
else:
    return struct.unpack("I", r)[0]
```

tiran commented on Modules/socketmodule.c:
```diff
@@ -2608,20 +2665,39 @@ sock_getsockopt(PySocketSockObject *s, PyObject *args)
-        socklen_t flagsize = sizeof flag;
+#ifdef AF_VSOCK
+        if (s->sock_family == AF_VSOCK) {
+            uint64_t vflag = 0;
```
I've been starring at this block for a minute until I realized that it handles ``uint64_t`` instead of ``int``. Please add a comment that this block that explains the difference.

tiran commented on Modules/socketmodule.c:
```diff
@@ -2608,20 +2665,39 @@ sock_getsockopt(PySocketSockObject *s, PyObject *args)
+                         (void *)&vflag, &flagsize);
+            if (res < 0)
+                return s->errorhandler();
+            return PyLong_FromLong(vflag);
```
It's an ``uint64_t``, you have to use ``PyLong_FromUnsignedLong()``.

tiran commented on Modules/socketmodule.h:
```diff
@@ -107,6 +107,15 @@ typedef int socklen_t;
+#ifdef HAVE_LINUX_VM_SOCKETS_H
+# include <linux/vm_sockets.h>
+#  ifndef AF_VSOCK
+#    define AF_VSOCK 40
```
Why is this necessary?

tiran reviewed (changes requested):
I added some comments.

You also have to add the vsock struct to ``typedef union sock_addr`` in ``socketmodule.h``.

Please use ``blurb`` to add a news entry, too.

caavery committed:
bpo-27584: Fixes for V6

Added news file.
socket.rst now reflects first Linux introduction of AF_VSOCK.
Fixed get_cid in test_socket.py.
Replaced PyLong_FromLong with PyLong_FromUnsignedLong in socketmodule.c
Got rid of extra AF_VSOCK #define.
Added sockaddr_vm to sock_addr.

caavery commented:
I didn't expect the Spanish Inquisition!

tiran commented on Modules/socketmodule.h:
```diff
@@ -107,6 +107,12 @@ typedef int socklen_t;
+#ifdef HAVE_LINUX_VM_SOCKETS_H
+# include <linux/vm_sockets.h>
+#else
+# undef AF_VSOCK
```
I don't think ``#undef AF_VSOCK`` is necessary. Please remove it.

caavery replied:
It needs to be there. In the case that vm_sockets.h is not available in include/linux which is the case on the last build bot run. Not all of the bots have the newer kernel.That way the code dependent on vm_sockets.h (AF_VSOCK) will not try to compile.

checking for linux/vm_sockets.h... no

This is also the way its done it at the top of socketmodule.h with AF_NETLINK

I also noticed AF_CAN was not adding the #undef AF_CAN in socketmodule.h so I renamed linux/can.h, ran configure and make and sure enough any code conditionally compiling under #ifdef AF_CAN in socketmodule.c failed.

tiran commented on Modules/socketmodule.h:
```diff
@@ -193,6 +199,9 @@ typedef union sock_addr {
 #ifdef HAVE_SOCKADDR_ALG
     struct sockaddr_alg alg;
 #endif
+#ifdef HAVE_LINUX_VM_SOCKETS_H
```
Please be consistent and use ``#ifdef AF_VSOCK`` in all places.

tiran reviewed (changes requested):
You are almost there!

caavery committed:
bpo-27584: Fixes for V7

Minor cleanup.

caavery commented:
OK there seems to be a problem with the build now. I'll address it next week. Thanks.

tiran commented:
You were right, either ``#undef AF_VSOCK`` is required or you have to replace all ``#ifdef AF_VSOCK`` with ``#ifdef HAVE_LINUX_VM_SOCKETS_H``. The address family and the VM socket struct are defined in different header files. ``AF_VSOCK`` can be present w/o ``linux/vm_sockets.h``.

caavery commented:
I didn't expect the Spanish Inquisition!

caavery committed:
bpo-27584: Fixes for V8

Put back #undef AF_VSOCK as it is  necessary when vm_sockets.h is not installed.

tiran reviewed (approved):
LGTM, thanks!
